File: Chapter 9/9.6-gslayered/ConfigureEnvironment/GLCoreProfileView.m
//
//  GLCoreProfileView.m
//  ConfigureEnvironment
//
//  Created by 陈杰 on 26/10/2017.
//  Copyright © 2017 陈杰. All rights reserved.
//

#import "GLCoreProfileView.h"
#import <OpenGL/OpenGL.h>
#import <GLKit/GLKit.h>
#import "TextureManager.h"
#import "TDModelManger.h"

@interface GLCoreProfileView()

@property (nonatomic, strong) NSTimer *lifeTimer;
@property (nonatomic, assign) CGFloat lifeDuration;
@property (nonatomic, assign) BOOL paused;

@property (atomic, assign) GLuint programGSLayers;
@property (atomic, assign) GLuint programShowLayers;

@property (atomic, assign) GLuint transformUBO;
@property (atomic, assign) GLuint layeredFBO;
@property (atomic, assign) GLuint colorArrayTexture;
@property (atomic, assign) GLuint depthArrayTexture;

@property (atomic, assign) GLint colorArrayTextureLocation;

@end

@implementation GLCoreProfileView
#pragma mark - lifecycle methods
- (instancetype)initWithCoder:(NSCoder *)decoder {
    NSOpenGLPixelFormatAttribute pixelFormatAttributes[] = {
        NSOpenGLPFAColorSize, 32,
        NSOpenGLPFADepthSize, 24,
        NSOpenGLPFAStencilSize, 8,
        NSOpenGLPFAAccelerated,
        NSOpenGLPFAOpenGLProfile, NSOpenGLProfileVersion4_1Core,
        0
    };
    NSOpenGLPixelFormat *pixelFormat = [[NSOpenGLPixelFormat alloc] initWithAttributes:pixelFormatAttributes];
    NSOpenGLContext *openGLContext = [[NSOpenGLContext alloc] initWithFormat:pixelFormat shareContext:nil];
    
    if (self = [super initWithCoder:decoder]) {
        [self setOpenGLContext:openGLContext];
        [self.openGLContext makeCurrentContext];
        GLint sync = 1;
        CGLSetParameter(CGLGetCurrentContext(), kCGLCPSwapInterval, &sync);
        
        _lifeTimer = [NSTimer scheduledTimerWithTimeInterval:0.02 target:self selector:@selector(lifeTimerUpdate) userInfo:nil repeats:YES];
        _paused = NO;
    
        __weak __typeof(self) weakself = self;
        [NSEvent addLocalMonitorForEventsMatchingMask:NSEventMaskKeyDown handler:^NSEvent * _Nullable(NSEvent * _Nonnull aEvent) {
            [weakself keyDown:aEvent];
            return aEvent;
        }];
    }
    return self;
}

- (void)keyDown:(NSEvent *)event {
    if ([event.characters isEqualToString:@"f"]) {
        
    } else if ([event.characters isEqualToString:@"d"]) {

    } else if ([event.characters isEqualToString:@"w"]) {
    } else if ([event.characters isEqualToString:@"p"]) {
        self.paused = !self.paused;
    } else if ([event.characters isEqualToString:@"+"]) {

    } else if ([event.characters isEqualToString:@"-"]) {

    }
    NSLog(@"%@",event.characters);
}

- (void)dealloc {
    [_lifeTimer invalidate];
    _lifeTimer = nil;

//    glDeleteBuffers(1, &_patchBuffer);
//    glDeleteBuffers(1, &_cageIndicesBuffer);
//    glDeleteVertexArrays(1, &_vertexArray);
//    glDeleteProgram(_tessProgram);
//    glDeleteProgram(_deawCpProgram);
}

- (void)prepareOpenGL {
    [super prepareOpenGL];

    NSLog(@"Version: %s", glGetString(GL_VERSION));
    NSLog(@"Renderer: %s", glGetString(GL_RENDERER));
    NSLog(@"Vendor: %s", glGetString(GL_VENDOR));
    NSLog(@"GLSL Version: %s", glGetString(GL_SHADING_LANGUAGE_VERSION));
    
    [self loadShaders];
    
    // 加载模型
    [[TDModelManger shareManager] loadObjectWithFileName:@"torus.sbm"];
    
    // 开启面剔除特性
    glEnable(GL_CULL_FACE);
    
    // 准备一致闭包变量缓存
    glGenBuffers(1, &_transformUBO);
    glBindBuffer(GL_UNIFORM_BUFFER, _transformUBO);
    glBufferData(GL_UNIFORM_BUFFER, sizeof(GLKMatrix4) * 17, NULL, GL_DYNAMIC_DRAW);
    
    // 准备FBO
    glGenFramebuffers(1, &_layeredFBO);
    glBindFramebuffer(GL_FRAMEBUFFER, _layeredFBO);
   
    glGenTextures(1, &_colorArrayTexture);
    glBindTexture(GL_TEXTURE_2D_ARRAY, _colorArrayTexture);
    size_t pixelDataSize = 512 * 512 * 4 * sizeof(GLubyte) * 16;
    void *pixelData = (void *)malloc(pixelDataSize);
    memset(pixelData, 0xFF, pixelDataSize);
    glTexImage3D(GL_TEXTURE_2D_ARRAY, 0, GL_RGBA8, 512, 512, 16, 0, GL_RGBA, GL_UNSIGNED_BYTE, pixelData);
    free(pixelData);
    glFramebufferTexture(GL_FRAMEBUFFER, GL_COLOR_ATTACHMENT0, self.colorArrayTexture, 0);
    glTexParameteri(GL_TEXTURE_2D_ARRAY, GL_TEXTURE_MIN_FILTER, GL_LINEAR);
    glTexParameteri(GL_TEXTURE_2D_ARRAY, GL_TEXTURE_MAG_FILTER, GL_LINEAR);

    size_t depthDataSize = 512 * 512 * 4 * sizeof(GLubyte) * 16;
    void *depthData = (void *)malloc(depthDataSize);
    memset(depthData, 0x00, depthDataSize);
    glGenTextures(1, &_depthArrayTexture);
    glBindTexture(GL_TEXTURE_2D_ARRAY, self.depthArrayTexture);
    glTexImage3D(GL_TEXTURE_2D_ARRAY, 0, GL_DEPTH_COMPONENT32F, 512, 512, 16, 0, GL_DEPTH_COMPONENT, GL_UNSIGNED_BYTE, depthData);
    free(depthData);
    glFramebufferTexture(GL_FRAMEBUFFER, GL_DEPTH_ATTACHMENT, self.depthArrayTexture, 0);

    // 检查FBO完整性
    GLenum status = glCheckFramebufferStatus(GL_FRAMEBUFFER);
    if (status == GL_FRAMEBUFFER_COMPLETE) {
        NSLog(@"ee");
    }
    
    // 设置绘制附件索引
    static const GLenum drawBuffers[] = {GL_COLOR_ATTACHMENT0};
    glDrawBuffers(1, drawBuffers);
}

- (void)reshape {
    [super reshape];
}

- (void)drawRect:(NSRect)dirtyRect {
    static double last_time = 0.0;
    static double total_time = 0.0;
    if (!_paused) {
        total_time += (_lifeDuration - last_time);
    }
    last_time = _lifeDuration;
    CGFloat progressTime = _lifeDuration * 0.5;
    
    // 为一致闭包变量赋值
    typedef struct TRANSFORM_BUFFER_S {
        GLKMatrix4 proj_matrix;
        GLKMatrix4 mv_matrix[16];
    } TRANSFORM_BUFFER;
    glBindBufferBase(GL_UNIFORM_BUFFER, 0, self.transformUBO);
    TRANSFORM_BUFFER *buffer = (TRANSFORM_BUFFER *)glMapBufferRange(GL_UNIFORM_BUFFER, 0, sizeof(TRANSFORM_BUFFER), GL_MAP_WRITE_BIT | GL_MAP_INVALIDATE_BUFFER_BIT);
    buffer->proj_matrix = GLKMatrix4MakePerspective(GLKMathDegreesToRadians(50.0f), 1.0f, 0.1f, 1000.0f);
    for (int i = 0; i < 16; i++) {
        float fi = (float)(i + 12) / 16.0f;
        GLKMatrix4 mv_Matrix = GLKMatrix4MakeTranslation(0.0f, 0.0f, -4.0f);
        mv_Matrix = GLKMatrix4Rotate(mv_Matrix, GLKMathDegreesToRadians(progressTime * 25.0f * fi), 0.0f, 0.0f, 1.0f);
        mv_Matrix = GLKMatrix4Rotate(mv_Matrix, GLKMathDegreesToRadians(progressTime * 30.0f * fi), 1.0f, 0.0f, 0.0f);
        buffer->mv_matrix[i] = mv_Matrix;
    }
    glUnmapBuffer(GL_UNIFORM_BUFFER);
    
    // 绘制纹理
    glBindFramebuffer(GL_FRAMEBUFFER, self.layeredFBO);
    glViewport(0, 0, 512, 512);
    static const GLfloat black[] = { 0.0f, 0.0f, 0.0f, 1.0f };
    glClearBufferfv(GL_COLOR, 0, black);
    static const GLfloat one = 1.0f;
    glClearBufferfv(GL_DEPTH, 0, &one);
    glEnable(GL_DEPTH_TEST);
    glDepthFunc(GL_LEQUAL);
    
    glUseProgram(self.programGSLayers);
    [[TDModelManger shareManager] render];
    
    // 将纹理绘制到屏幕上
    glBindFramebuffer(GL_FRAMEBUFFER, 0);
    glDrawBuffer(GL_BACK);
    static const GLfloat gray[] =  { 0.1f, 0.1f, 0.1f, 1.0f };
    glClearBufferfv(GL_COLOR, 0, gray);
    glClearBufferfv(GL_DEPTH, 0, &one);
    NSRect bounds = [self bounds];
    glViewport(0, 0, NSWidth(bounds), NSHeight(bounds));

    glUseProgram(self.programShowLayers);
    glActiveTexture(GL_TEXTURE0);
    glBindTexture(GL_TEXTURE_2D_ARRAY, self.colorArrayTexture);
    glUniform1i(self.colorArrayTextureLocation, 0);
    glDisable(GL_DEPTH_TEST);
    glBindVertexArray([TDModelManger shareManager].vao);
    glDrawArraysInstanced(GL_TRIANGLE_FAN, 0, 4, 16);
    
    glBindTexture(GL_TEXTURE_2D_ARRAY, 0);
    glFlush();
}

#pragma mark - private methods
- (BOOL)loadShaders {
    // programGSLayers
    GLuint gslayersShaders[3];
    GLushort gslayerShaderTypes[3] = {GL_VERTEX_SHADER, GL_GEOMETRY_SHADER, GL_FRAGMENT_SHADER};
    NSString *gslayersVSShaderPath = [[NSBundle mainBundle] pathForResource:@"gslayersVS" ofType:@"glsl"];
    NSString *gslayersGSShaderPath = [[NSBundle mainBundle] pathForResource:@"gslayersGS" ofType:@"glsl"];
    NSString *gslayersFSShaderPath = [[NSBundle mainBundle] pathForResource:@"gslayersFS" ofType:@"glsl"];
    NSArray *gslayerShaderPathes = @[gslayersVSShaderPath, gslayersGSShaderPath, gslayersFSShaderPath];
    
    for (int i = 0; i < 3; i++) {
        if (![self compileShader:&gslayersShaders[i] type:gslayerShaderTypes[i] filePath:gslayerShaderPathes[i]]) {
            NSLog(@"Failed to compile shader at %@", gslayerShaderPathes[i]);
        }
    }
    
    self.programGSLayers = glCreateProgram();
    for (int i = 0; i < 3; i++) {
        glAttachShader(self.programGSLayers, gslayersShaders[i]);
        glDeleteShader(gslayersShaders[i]);
        gslayersShaders[i] = 0;
    }
    
    if (![self linkProgram:self.programGSLayers]) {
        NSLog(@"Failed to link programGSLayers");
        if (self.programGSLayers != 0) {
            glDeleteProgram(self.programGSLayers);
            self.programGSLayers = 0;
        }
        return NO;
    }
    
    // programShowLayers
    GLuint showlayersShaders[2];
    GLushort showlayersShaderTypes[2] = {GL_VERTEX_SHADER, GL_FRAGMENT_SHADER};
    NSString *showlayersVSShaderPath = [[NSBundle mainBundle] pathForResource:@"showlayersVS" ofType:@"glsl"];
    NSString *showlayersFSShaderPath = [[NSBundle mainBundle] pathForResource:@"showlayersFS" ofType:@"glsl"];
    NSArray *showlayersShaderPathes = @[showlayersVSShaderPath, showlayersFSShaderPath];
    
    for (int i = 0; i < 2; i++) {
        if (![self compileShader:&showlayersShaders[i] type:showlayersShaderTypes[i] filePath:showlayersShaderPathes[i]]) {
            NSLog(@"Failed to compile shader at %@", showlayersShaderPathes[i]);
        }
    }
    
    self.programShowLayers = glCreateProgram();
    for (int i = 0; i < 3; i++) {
        glAttachShader(self.programShowLayers, showlayersShaders[i]);
        glDeleteShader(showlayersShaders[i]);
        showlayersShaders[i] = 0;
    }
    
    if (![self linkProgram:self.programShowLayers]) {
        NSLog(@"Failed to link programGSLayers");
        if (self.programShowLayers != 0) {
            glDeleteProgram(self.programShowLayers);
            self.programShowLayers = 0;
        }
        return NO;
    }

    self.colorArrayTextureLocation = glGetUniformLocation(self.programShowLayers, "tex");
    return YES;
}

- (BOOL)compileShader:(GLuint *)shader type:(GLenum)type filePath:(NSString *)path {
    const GLchar *shaderSource = [NSString stringWithContentsOfFile:path encoding:NSUTF8StringEncoding error:nil].UTF8String;
    *shader = glCreateShader(type);
    glShaderSource(*shader, 1, &shaderSource, nil);
    glCompileShader(*shader);
    
    GLint status = 0;
    glGetShaderiv(*shader, GL_COMPILE_STATUS, &status);
    if (status == 0) {
        GLint logLen = 0;
        glGetShaderiv(*shader, GL_INFO_LOG_LENGTH, &logLen);
        GLchar *infoLog = malloc(sizeof(char) * logLen);
        glGetShaderInfoLog(*shader, logLen, NULL, infoLog);
        NSLog(@"Shader at: %@", path);
        fprintf(stderr, "Info Log: %s\n", infoLog);
        
        glDeleteShader(*shader);
        return NO;
    }
    return YES;
}

- (BOOL)linkProgram:(GLuint)program {
    glLinkProgram(program);
    GLint status = 0;
    glGetProgramiv(program, GL_LINK_STATUS, &status);
    if (status == 0) {
        GLint logLen = 0;
        glGetProgramiv(program, GL_INFO_LOG_LENGTH, &logLen);
        GLchar *infoLog = malloc(sizeof(char) * logLen);
        glGetProgramInfoLog(program, logLen, NULL, infoLog);
        fprintf(stderr, "Prog Info Log: %s\n", infoLog);
        return NO;
    }
    return YES;
}

#pragma mark - listening methods
- (void)lifeTimerUpdate {
    _lifeDuration += _lifeTimer.timeInterval;
    // _lifeDuration为程序运行时间
    [self drawRect:self.bounds];
}

#pragma mark - accessor methods

@end

File: Chapter 12/ToonShading/Source/GLCoreProfileView.swift
//
//  GLCoreProfileView.swift
//  ToonShading
//
//  Created by chenjie on 2019/12/3.
//  Copyright © 2019 chenjie. All rights reserved.
//

import GLKit

fileprivate struct UniformsLocation {
    var mv_matrix: GLint = 0
    var proj_matrix: GLint = 0
    var tex_toon: GLint = 0
    var light_pos: GLint = 0
}

class GLCoreProfileView: NSOpenGLView {
    //MARK:- Public properties
    // 除去暂停后实际渲染的时间
    var renderDuration: TimeInterval = 0
    
    //MARK:- Private Properties
    private var glProgram: GLuint = 0
    private var uniformsLocation = UniformsLocation()
    private var toonTexture: GLuint = 0

    //MARK:- Life Cycles
    override init?(frame frameRect: NSRect, pixelFormat format: NSOpenGLPixelFormat?) {
        super.init(frame: frameRect, pixelFormat: format)
        // 禁用Retina屏幕显示
        self.wantsBestResolutionOpenGLSurface = false

        self.prepareOpenGLContex()
    }
    
    required init?(coder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
    
    deinit {
        glDeleteProgram(self.glProgram)
        glDeleteTextures(1, &self.toonTexture)
    }
    
    //MARK:- Private Methods: General
    override func reshape() {
        super.reshape()
        glViewport(0, 0, GLsizei(NSWidth(self.frame)), GLsizei(NSHeight(self.frame)))
    }
    
    //MARK:- Private Methods: Render
    private func prepareOpenGLContex() {
        let pixelFormatAttributes = [UInt32(NSOpenGLPFAColorSize), 32,
                                     UInt32(NSOpenGLPFADepthSize), 24,
                                     UInt32(NSOpenGLPFAStencilSize), 8,
                                     UInt32(NSOpenGLPFAAccelerated),
                                     UInt32(NSOpenGLPFAOpenGLProfile), UInt32(NSOpenGLProfileVersion4_1Core),
                                     0]
        guard let pixelFormat = NSOpenGLPixelFormat(attributes: pixelFormatAttributes) else { return }
        let currentOpenGLContext = NSOpenGLContext(format: pixelFormat, share: nil)
        self.openGLContext = currentOpenGLContext
        self.openGLContext?.makeCurrentContext()
        // 设置缓存交换频率和屏幕刷新同步
        var swapInterval = GLint(1)
        CGLSetParameter(CGLGetCurrentContext()!, kCGLCPSwapInterval, &swapInterval)
    }

    override func prepareOpenGL() {
        super.prepareOpenGL()
        print("Version: \(String(cString: glGetString(uint32(GL_VERSION))))")
        print("Renderer: \(String(cString: glGetString(uint32(GL_RENDERER))))")
        print("Vendor: \(String(cString: glGetString(GLenum(GL_VENDOR))))")
        print("GLSL Version: \(String(cString: glGetString(GLenum(GL_SHADING_LANGUAGE_VERSION))))")
        
        // 1. 准备所有的OpenGL程序
        if !self.prepareOpenGLProgram() {
            return
        }
        
        // 2. 加载纹理
        glGenTextures(1, &self.toonTexture)
        glActiveTexture(GLenum(GL_TEXTURE0))
        glBindTexture(GLenum(GL_TEXTURE_1D), self.toonTexture)
        var data: [uint8] = [0x44, 0x00, 0x00, 0x00,
                             0x88, 0x00, 0x00, 0x00,
                             0xCC, 0x00, 0x00, 0x00,
                             0xFF, 0x00, 0x00, 0x00]
        glTexImage1D(GLenum(GL_TEXTURE_1D), 0, GL_RGB8, 4, 0, GLenum(GL_RGBA), GLenum(GL_UNSIGNED_BYTE), &data)
        glTexParameteri(GLenum(GL_TEXTURE_1D), GLenum(GL_TEXTURE_MAG_FILTER), GL_NEAREST)
        glTexParameteri(GLenum(GL_TEXTURE_1D), GLenum(GL_TEXTURE_MIN_FILTER), GL_NEAREST)
        glTexParameteri(GLenum(GL_TEXTURE_1D), GLenum(GL_TEXTURE_WRAP_S), GL_CLAMP_TO_EDGE)
        glBindTexture(GLenum(GL_TEXTURE_1D), 0)
        
        
        // 3. 加载模型
        let modelLoaded = ModelManager.shareManager.loadObject(fileName: "torus_nrms_tc.sbm")
        if !modelLoaded {
            print("Load model failed.")
            return
        }

        // 4. 开启面剔除和深度测试
        glEnable(GLenum(GL_CULL_FACE))
        glEnable(GLenum(GL_DEPTH_TEST))
        glDepthFunc(GLenum(GL_LEQUAL))
    }

    private func prepareOpenGLProgram() -> Bool {
        // 1. 准备阴影制作的OpenGL程序
        let shaders = ["VertexShader" : GLenum(GL_VERTEX_SHADER),
                       "FragmentShader" : GLenum(GL_FRAGMENT_SHADER)]
        let successed = self.prepareGLProgram(&self.glProgram, shaders: shaders)
        if !successed {
            return false
        }
        
        // 2. 获取统一变量的位置
        self.uniformsLocation.mv_matrix = glGetUniformLocation(self.glProgram, "mv_matrix")
        self.uniformsLocation.proj_matrix = glGetUniformLocation(self.glProgram, "proj_matrix")
        self.uniformsLocation.tex_toon = glGetUniformLocation(self.glProgram, "tex_toon")
        self.uniformsLocation.light_pos = glGetUniformLocation(self.glProgram, "light_pos")
        return true
    }
            
    private func prepareGLProgram(_ glProgram: inout GLuint, shaders: [String : GLenum]) -> Bool {
        if shaders.count == 0 {
            print("No available shader.")
            return false
        }
        
        // 1. 创建GL程序
        glProgram = glCreateProgram()

        for (_, shaderInfo) in shaders.enumerated() {
            // 2. 获取着色器的源码
            guard let shaderPath = Bundle.main.path(forResource: shaderInfo.key, ofType: "glsl") else {
                print("Can not find shader file at " + shaderInfo.key)
                return false
            }
            
            // 3. 创建着色器
            let shader = glCreateShader(shaderInfo.value)
            
            // 4. 编译着色器
            if !self.compileShader(shader, filePath: shaderPath) {
                glDeleteShader(shader)
                return false
            }

            // 5. 将着色器附着至OpenGL程序
            glAttachShader(glProgram, shader)
            glDeleteShader(shader)
        }
        
        // 6. 编译GL程序
        if !self.linkProgram(program: glProgram) {
            print("Failed to link program")
            glDeleteProgram(glProgram)
            glProgram = 0
            return false
        }
        return true
    }
        
    private func compileShader(_ shader: GLuint, filePath: String) -> Bool {
        // 1. 为着色器填充源码
        var shaderSource: UnsafePointer<GLchar>?
        do {
            shaderSource = try NSString(contentsOfFile: filePath, encoding: String.Encoding.utf8.rawValue).utf8String
        } catch {
            print("Error when compile shader")
        }
        glShaderSource(shader, 1, &shaderSource, nil)
        // 2. 编译着色器
        glCompileShader(shader)
        
        // 3. 查询着色器编译错误日志
        var status: GLint = 0
        glGetShaderiv(shader, GLenum(GL_COMPILE_STATUS), &status)
        if status == 0 {
            var logLength: GLint = 0
            glGetShaderiv(shader, GLenum(GL_INFO_LOG_LENGTH), &logLength)
            var infoLogChars = [GLchar](repeating: 0, count: Int(logLength))
            glGetShaderInfoLog(shader, logLength, nil, &infoLogChars)
            let infoLogString = String(utf8String: infoLogChars) ?? ""
            print("Compile Shader Failed at \n" + filePath + "\nWith Log\n" + infoLogString)
            return false
        }
        return true
    }

    func linkProgram(program: GLuint) -> Bool {
        glLinkProgram(program)
        var status: GLint = 0
        glGetProgramiv(program, GLenum(GL_LINK_STATUS), &status)
        if status == 0 {
            var logLength: GLint = 0
            glGetProgramiv(program, GLenum(GL_INFO_LOG_LENGTH), &logLength)
            var logChars = [GLchar](repeating: 0, count: Int(logLength))
            glGetProgramInfoLog(program, GLsizei(logLength), nil, &logChars)
            let logString = String(utf8String: logChars) ?? ""
            print("Compile program failed with log \n" + logString)
            return false
        }
        
        return true
    }
        
    override func draw(_ dirtyRect: NSRect) {
        // 1. 清空颜色和深度缓存
        var blackColor: [GLfloat] = [0.1, 0.1, 0.1, 1]
        glClearBufferfv(GLenum(GL_COLOR), 0, &blackColor)
        var defaultDepth: GLfloat = 1
        glClearBufferfv(GLenum(GL_DEPTH), 0, &defaultDepth)
        
        // 2. 设置OpenGL程序
        glUseProgram(self.glProgram)
        
        // 3. 为统一变量赋值
        glActiveTexture(GLenum(GL_TEXTURE0))
        glBindTexture(GLenum(GL_TEXTURE_1D), self.toonTexture)
        glUniform1i(self.uniformsLocation.tex_toon, 0)
        
        let translateMatrix = GLKMatrix4MakeTranslation(0, 0, -3)
        let yRotateMatrix = GLKMatrix4MakeYRotation(GLKMathDegreesToRadians(Float(self.renderDuration * 13.75)))
        let zRotateMatrix = GLKMatrix4MakeZRotation(GLKMathDegreesToRadians(Float(self.renderDuration * 7.75)))
        let xRotateMatrix = GLKMatrix4MakeXRotation(GLKMathDegreesToRadians(Float(self.renderDuration * 15.3)))
        var mv_matrix = GLKMatrix4Multiply(translateMatrix, GLKMatrix4Multiply(yRotateMatrix, GLKMatrix4Multiply(zRotateMatrix, xRotateMatrix)))
        withUnsafePointer(to: &mv_matrix.m) {
            $0.withMemoryRebound(to: GLfloat.self, capacity: 16) {
                glUniformMatrix4fv(self.uniformsLocation.mv_matrix, 1, GLboolean(GL_FALSE), $0)
            }
        }
        
        var proj_matrix = GLKMatrix4MakePerspective(GLKMathDegreesToRadians(60),
                                                    Float(NSWidth(self.frame) / NSHeight(self.frame)),
                                                    0.1, 1000)
        withUnsafePointer(to: &proj_matrix.m) {
            $0.withMemoryRebound(to: GLfloat.self, capacity: 16) {
                glUniformMatrix4fv(self.uniformsLocation.proj_matrix, 1, GLboolean(GL_FALSE), $0)
            }
        }
        
        // 4. 绘制场景
        ModelManager.shareManager.render()
        
        // 5. 清空OpenGL指令缓存，使渲染引擎尽快执行已经发布的渲染指令，同时发布窗口缓存的数据已经做好合并进入系统桌面缓存的准备
        glFlush()
        
        // 6. 复原纹理顶点绑定状态
        glBindTexture(GLenum(GL_TEXTURE_1D), 0)
    }
}
